### PR TITLE
Support Go 1.13 and Go 1.14 both

### PR DESCRIPTION
**Pull request by visualfc**
Build on Go 1.13 / Go 1.14 and one build support working on Go 1.12 Go 1.13 Go 1.14 three version.

```
cd goproj/src
git clone https://github.com/visualfc/gopherjs github.com/gopherjs/gopherjs
cd github.com/gopherjs/gopherjs
git checkout go1.13-dev
go install -v
```
1. This version build and install on Go 1.13 or Go 1.14. And one build support working on Go 1.12 1.13 1.14 three version and dynamic select internal compiler/natives/src.
2. This version build and install on Go 1.12 and working on Go 1.12 only.
3. if change  install  go version, please use -a flags to force rebuild packages.
`gopherjs build -a -v` or `gopherjs test -a -v`

* add internal/reflectlite for Go 1.13  Go 1.14
* update syscall/js API for Go 1.13 Go 1.14 changes
* support Go Module, build go project check go.mod ( use go env)
* add -a (--force) flags to force rebuild packages
* check installed go version for build ReleaseTags, support working Go 1.12 Go 1.13 Go 1.14 three version

### CODE REVIEW

**Comment by dmitshur** [comment]
CI is failing because the `circle.yml` file in this PR doesn't include the fix for issue #937 that was applied in commit 1852f3ae295149668e1c8d407098e9ed35df5513. @visualfc Can you include it?

**Comment by visualfc** [comment]
> CI is failing because the `circle.yml` file in this PR doesn't include the fix for issue #937 that was applied in commit [1852f3a](https://github.com/gopherjs/gopherjs/commit/1852f3ae295149668e1c8d407098e9ed35df5513). @visualfc Can you include it?

update, but tests failed for CI.

**Comment by dmitshur** [comment]
It's failing because the `diff -u <(echo -n) <(go list ./compiler/natives/src/...)` step is reporting a non-zero diff. You should add a `// +build js` constraint to the newly added files in that directory.

**Comment by visualfc** [comment]
@dmitshur update for CI check.

**Comment by dmitshur** [comment]
Congratulations, and thank you for getting the PR to the point of passing CI with go1.13.6, that is a fantastic accomplishment! It brings us closer to a future GopherJS 1.13 or 1.14 version.

We need to make some decisions next in order to proceed with getting this merged into `master`.

Generally, the 1.n -> 1.n+1 PRs are large and complicated, because there are many things that need to be accounted for in order to upgrade GopherJS to support the next release of Go. As a result, we don't want to increase the scope beyond the minimum viable changes.

This PR currently takes on many things, including:

- adding a go.mod file to the GopherJS project itself
- support for multiple Go versions at the same time (1.14, 1.13, and 1.12)
- support for Go modules in packages being built, if I understand correctly

These are good changes, but we need to review them all, and it's easier to do it separately, rather than as part of one very large change.

Historically, GopherJS version 1.n-x means it supports exactly Go 1.n and none of the earlier or later versions. This has limitations, but keeps things simpler, which is especially important today given our limited resources (see #894).

I suggest we agree on a single Go version to target (either 1.13 or 1.14), then put some of the current development work here on the side, and create a simpler version of this change that only updates the support for that Go version. Then we can start reviewing that.

How does this sound @visualfc?

**Comment by visualfc** [comment]
@dmitshur I can rewrite code submission

**Comment by visualfc** [comment]
@dmitshur OK, I can rewrite code submission on go1.13 branch based.
The master branch set GOPHERJS_GOROOT environment variable to keep go1.12 version.  But GOPHERJS_GOROOT may not be needed need,  the multiple Go version support is simple and effective. 
The build system select current installed Go version to make ReleaseTags. And compiler/natives/src packages use +build version to auto select current Go version support build.
The  native packages diffrence between Go 1.12 and Go 1.13 is internal/reflectlite sync reflect syscall/js
The  native packages diffrence between Go 1.13 and Go 1.14 is packages syscall/js math

**Comment by visualfc** [comment]
@dmitshur which version branch ? create simple go 1.13 target version build.

**Comment by trende-jp** [comment]
The current 1.13 branch has few changes from the master.  If I were you I would work off of the existing 1.13 and only make changes specific to 1.13.

**Comment by benma** [comment]
What's the status? Would be great to be able to use Go 1.14 :heart:

**Comment by pjebs** [comment]
Any news?

**Comment by chenrui333** [comment]
Any updates on this PR? Thanks!

**Comment by nevkontakte** [comment]
Hi @dmitshur, @visualfc and everyone else!

I was suddenly made aware that Go 1.15 is practically around the corner, so soon enough we'll have to begin worrying about supporting that :) Addressing Dmitri's question in https://github.com/gopherjs/gopherjs/pull/964#issuecomment-587057441 on how to proceed with version branches, I'd like to suggest skipping 1.13 entirely and merging this PR as Go 1.14 support.

This PR is pretty large, though, so I would like to offer my help reviewing it, if would help moving along and everyone is ok with that.

**Comment by nevkontakte** [comment]
Since I received no objections, I'll start reviewing this PR in my spare time. It may take a while due to the size of the change, though.

@visualfc do I understand it correctly that https://github.com/gopherjs/gopherjs/pull/971 is a stcrict subset of this one?

File: compiler/gopherjspkg/fs.go [hunk 1]
```diff
@@ -23,6 +24,12 @@ var FS = filter.Keep(
 )
 
 func importPathToDir(importPath string) string {
+	for _, src := range build.Default.SrcDirs() {
+		dir := filepath.Join(src, importPath)
+		if _, err := os.Stat(dir); err == nil {
+			return dir
+		}
+	}
```
Comment (nevkontakte) [inline compiler/gopherjspkg/fs.go:32]: Why is this needed? I'd expect build.Import() to find packages in standard locations. Same below.

File: compiler/natives/src/internal/reflectlite/reflectlite.go [hunk 2]
```diff
@@ -0,0 +1,1048 @@
+// +build js
```
Comment (nevkontakte) [inline compiler/natives/src/internal/reflectlite/reflectlite.go:1]: So far I tool only cursory look at `reflectlite` code. I assume it is mostly derived from GopherJS version of `reflect`?

Is it possible to somehow reduce the amount of code duplication between the two?

**Comment by nevkontakte** [comment]
I'm getting a feeling like I'm talking to the void here :-/ @hajimehoshi @dmitshur @visualfc  Is there any interest in merging this PR at all?

**Comment by pjebs** [comment]
You are correct in your deduction. The project is dead and the maintainers don't care. I have an ignored PR too but I've known it's been dead for 1+ years.

It's really time for a hard fork for people such as you and I that actually use it in prod. They'll never transfer ownership of repo. I am the owner of github.com/rocketlaunchr/react repo which I use in prod and which relies on gopherjs.

**Comment by inliquid** [comment]
> I'm getting a feeling like I'm talking to the void here :-/ @hajimehoshi @dmitshur @visualfc Is there any interest in merging this PR at all? Any way I can help advancing this?
Seems pretty much so.

**Comment by nevkontakte** [comment]
To be clear, I recognize that GopherJS maintainers may be busy with other things (work, life, etc) and have little or no time to dedicate to the project at this time, but I'd like to hear at least some statement of intentions related to this PR. A hard fork would be very unfortunate and I'd very much prefer to avoid it.

Cheers!

**Comment by hajimehoshi** [comment]
Sorry but currently I don't have bandwidth to care this project. Moreover, I'm not familiar with the internal details of GopherJS compared to @dmitshur and the other maintainers.

> It's really time for a hard fork for people such as you and I that actually use it in prod.

Actually there are forks like https://github.com/myitcv/gopherjs, which has already merged some PRs in this project.

File: compiler/natives/src/internal/reflectlite/reflectlite.go (hunk 2 above)
Comment (visualfc) [reply compiler/natives/src/internal/reflectlite/reflectlite.go:1]: reflectlite clone from goperhjs reflect and comment unused code.

**Comment by visualfc** [comment]
> I'm getting a feeling like I'm talking to the void here :-/ @hajimehoshi @dmitshur @visualfc Is there any interest in merging this PR at all? Any way I can help advancing this?

this pr work for go1.14, but not support go1.15. seem suport go1.15 is hard.

**Comment by aubelsb2** [comment]
@hajimehoshi would it be possible to nominate a successor fork in the readme?

**Comment by benma** [comment]
I also use this in production. @dmitshur can you please bring some clarity in here? :pray:

**Comment by hajimehoshi** [comment]
@aubelsb2  I'm fine to add links to other forks, though my concern is whether this could be fare or not. What do you think, @dmitshur?

CC @myitcv

**Comment by flimzy** [comment]
> @hajimehoshi would it be possible to nominate a successor fork in the readme?

I think it would be better to see if @neelance and @dmitshur would be willing to promote one or two new owners for this project. WDYT?

**Comment by neelance** [comment]
> As a general comment, this PR attempts to support 3 Go versions simultaneously, but this seems to add a noticeable amount of complexity. Given that the amount of time GopherJS maintainers' seem to have available these days, it seems to set the project up for increased burden in future. I'd suggest dropping support for Go 1.12 and 1.13, and going with just Go 1.14. Given that 1.15 is coming soon, the value of keeping support for the older versions is quickly diminishing.

I fully agree with this. The project is already quite complex in what it does, it is better to avoid any additional complexity.

@nevkontakte I haven't talked to @dmitshur yet, but would you be open to being a maintainer of this project? You seem to be invested in GopherJS and I appreciate your code review.

**Comment by nevkontakte** [comment]
@neelance thanks for the offer. The amount of spare time I have is also limited, but I do indeed have a few projects in production that rely on GopherJS. If there's no better volunteer available, I'm happy to join :)

**Comment by nevkontakte** [comment]
Any news? I have multiple reasons to prefer GopherJS over Go/WASM, but at some point being able to use an up to date version will outweigh the advantages of GopherJS and the headache of migration :-(

**Comment by pjebs** [comment]
If someone had forks, I'm happy to migrate over

**Comment by inliquid** [comment]
@nevkontakte this is from Slack:
изображение

**Comment by flimzy** [comment]
:( Sounds like a fork is in order, then.

**Comment by neelance** [comment]
Yes, feel free to create a fork. If it is well maintained for a while, then we can endorse it or maybe even consider a merge at some point. But first Dmitri and I would like to see that the maintenance situation turns out better than what we are currently able to provide. :)

**Comment by pjebs** [comment]
what should it be called? `jsgopher` or `gojs`?

**Comment by flimzy** [comment]
> what should it be called? `jsgopher` or `gojs`?

I suggest we either open a new issue, and/or open a slack conversation about this, so we don't clutter this PR, and so we have a canonical place to point people for posterity. WDYT?

**Comment by nevkontakte** [comment]
@inliquid thanks for bringing the clarity.

@flimzy I suppose https://github.com/gopherjs/gopherjs/issues/894 is an appropriate place for this kind of discussion. Let's continue it there, shall we?

**Comment by visualfc** [comment]
support Go1.15 https://github.com/visualfc/gopherjs/tree/go1.15-dev

**Comment by pjebs** [comment]
@visualfc Can you merge it to master in your fork so I can use it? I think it's time to adopt yours as the "official" gopherjs.
If you want, you can create an organization. Perhaps called: "gojs". And then transfer the repo under that organization.

Thanks for the amazing work.

**Comment by visualfc** [comment]
https://github.com/goplusjs/gopherjs

**Comment by pjebs** [comment]
Does it pass all the go unit tests?

**Comment by visualfc** [comment]
> Does it pass all the go unit tests?

test pass on macOS Go1.12~Go1.15

**Comment by flimzy** [comment]
> @neelance thanks for the offer. The amount of spare time I have is also limited, but I do indeed have a few projects in production that rely on GopherJS. If there's no better volunteer available, I'm happy to join :)

@nevkontakte, are you still interested in helping to maintain this project? If so, would you mind adding your name to the document mentioned here? https://github.com/gopherjs/gopherjs/pull/964#issuecomment-769318251  And if not, that's of course perfectly understandable.

**Comment by nevkontakte** [comment]
@flimzy I already have (the second entry) 😃

**Comment by flimzy** [comment]
> @flimzy I already have (the second entry)

Ha! Sorry I didn't make the connection. And happy you've expressed interest :)

**Comment by flimzy** [comment]
@visualfc I should have pinged you, too. I know you're making pretty strong headway on your own fork now, so may not be interested in GopherJS any longer, but I want to make sure you have a chance to voice your interest, if you want. https://github.com/gopherjs/gopherjs/pull/964#issuecomment-769318251

I'm going to try to schedule a call with those interested in helping, to discuss next steps, and if you are interested, of course I'd love to involve you.

**Comment by visualfc** [comment]
> @visualfc I should have pinged you, too. I know you're making pretty strong headway on your own fork now, so may not be interested in GopherJS any longer, but I want to make sure you have a chance to voice your interest, if you want. [#964 (comment)](https://github.com/gopherjs/gopherjs/pull/964#issuecomment-769318251)
> 
> I'm going to try to schedule a call with those interested in helping, to discuss next steps, and if you are interested, of course I'd love to involve you.

I hope GopherJS can support more go versions. my fork and experiment can support go1.16

**Comment by flimzy** [comment]
> I hope GopherJS can support more go versions. my fork and experiment can support go1.16rc1.

Does this mean you're not interested in participating in the GopherJS planning call?

**Comment by visualfc** [comment]
> > I hope GopherJS can support more go versions. my fork and experiment can support go1.16rc1.
> 
> Does this mean you're not interested in participating in the GopherJS planning call?

If gopherjs has plans to review PR, I will submit PR again.

**Comment by flimzy** [comment]
> If gopherjs has plans to review PR, I will submit PR again.

That's clear. Thank you for your response. Hopefully we can start reviewing PRs again soon!

**Comment by visualfc** [comment]
https://github.com/goplusjs/gopherjs/releases/tag/v1.1.8

**Comment by pjebs** [comment]
I want to make my https://github.com/rocketlaunchr/react package compatible with gopherjs and your package @visualfc . Do you know how I can do it?

**Comment by visualfc** [comment]
> I want to make my https://github.com/rocketlaunchr/react package compatible with gopherjs and your package @visualfc . Do you know how I can do it?

project keep stay import "github.com/gopherjs/gopherjs/js" or use "syscall/js" for compatible wasm
`go get -u github.com/goplusjs/gopherjs` for install gopherjs

**Comment by visualfc** [comment]
https://github.com/goplusjs/gopherjs now support Go 1.12~1.16

Go1.12~1.15 test actions on macOS, Go1.16 test action on Ubuntu.

[Go1.12](https://github.com/goplusjs/gopherjs/actions?query=workflow%3AGo1.12)
[Go1.13](https://github.com/goplusjs/gopherjs/actions?query=workflow%3AGo1.13)
[Go1.14](https://github.com/goplusjs/gopherjs/actions?query=workflow%3AGo1.14)
[Go1.15](https://github.com/goplusjs/gopherjs/actions?query=workflow%3AGo1.15)
[Go1.16](https://github.com/goplusjs/gopherjs/actions?query=workflow%3AGo1.16)

**Comment by pjebs** [comment]
@visualfc I made a PR in your repo.

**Comment by flimzy** [comment]
Now that #1015 has been merged, GopherJS is up to date with the latest upstream Go, so I'm going to close this PR.
